acp/acp.go

```diff
@@ -0,0 +1,113 @@
+/*
+Package acp utilizes the sourcehub acp module to bring the functionality to defradb, this package also helps
+avoid the leakage of direct sourcehub references through out the code base, and eases in swapping
+between local embedded use case and a more global on sourcehub use case.
+*/
+
+package acp
+
+import (
+	"context"
+
+	"github.com/sourcenetwork/immutable"
+
+	"github.com/sourcenetwork/defradb/logging"
+)
+
+var (
+	log = logging.MustNewLogger("acp")
+
+	NoACPModule = immutable.None[ACPModule]()
+)
+
+// ACPModule is the interface to all types of access control modules there might exist.
+//
+// An embedded local implementation might restrict the implementation to not make remote calls.
+//
+// A remote implementation might allow remote calls to access the acp module in the cloud.
+// Note: sourcehub acp module will likely need the remote implementation.
+type ACPModule interface {
+
+	// Start initializes the acp module, with persistence data stored under the given folder.
+	//
+	// If previously started in the same path, then recovers / reloads (unless used in-memory).
+	Start(context.Context, string) error
+
+	// Close closes the resources in use by the acp module.
+	Close() error
+
+	// AddPolicy attempts to add the given policy. Detects the format of the policy automatically
+	// by assuming YAML format if JSON validation fails. Upon success a policyID is returned,
+	// otherwise returns error.
+	//
+	// A policy can not be added without a valid creator.
+	AddPolicy(ctx context.Context, creator string, policy string) (string, error)
+
+	// ValidateResourceExistsOnValidDPI consumes a policyID and resource to validate that the
+	// the following conditions are met:
+	// - policyID is a valid argument.
+	// - resource name is a valid argument.
+	// - policyID corresponds to a valid policy that is registered on acp module.
+	// - resource name exists on that registered policy.
+	// - resource is DPI compliant (more information on DPI below).
+	//
+	// An error is returned upon validation failure (i.e. if above rules not met).
+	//
+	// To be a DPI compliant resource, these rules MUST be met:
+	// - The resource must exist.
+	// - The resource must atleast have the required registerer relation under `relations`.
+	// - The resource must atleast have all the required permissions under `permissions`.
+	// - Every required permission must have the required registerer relation in `expr`.
+	// - The required registerer relation must be the leading (first) relation in `expr`.
+	//
+	// Learn more about the DefraDB Policy Interface [DPI](/acp/DPI.md)
+	// Learn more about DefraDB ACP Terminologies [TERMINOLOGY](/acp/TERMINOLOGY.md)
+	ValidateResourceExistsOnValidDPI(ctx context.Context, policyID, resource string) error
+
+	// RegisterDocCreation registers the document (object) to have access control.
+	// No error is returned upon successful registering of a document.
+	//
+	// Note(s):
+	// - This should be used upon document creation only.
```

Comment:
question: Why only on document creation?

This seems extra odd given that there is no transaction in the params, suggesting that it must always be done after document creation anyway (unless premtively generating the doc id - is that permitted?).

question: What happens if you call this twice for the same document?

question: What happens if you call this with a doc id that doesn't exist?

Reply:
> question: Why only on document creation?

Because that is when access control of documents belonging to permissioned schema starts. You can't / should not register the document after it has already been registered (infact even upon soft deletion we don't want to unregister either, even on hard delete john made a case to why we might still want to keep it registered).


> This seems extra odd given that there is no transaction in the params, suggesting that it must always be done after document creation anyway (unless premtively generating the doc id - is that permitted?).

It must only be done for documents that are NOT public. 

Are you asking if a public document (unregistered document) can be registered later? That will be added in a separate feature, and another feature that will allow users to share documents.

> question: What happens if you call this twice for the same document?

Sourcehub will return an error saying the object was already registered.

> question: What happens if you call this with a doc id that doesn't exist?

ACP has no knowledge of docID valid or not, is why all the steps are done after document is created, if you look at when it is called.

Reply:
> You can't / should not register the document after it has already been registered

I'm not talking about multiple registrations here, I'm asking about document creation.  I.e. create doc, wait, update doc, wait, etc, etc, register doc.

> Are you asking if a public document (unregistered document) can be registered later?

All documents are unregistered at somepoint with the current flow - as the document must exist before it is registered, and if it is not registered and does exist, it must be unregistered no?

> ACP has no knowledge of docID valid or not, is why all the steps are done after document is created, if you look at when it is called.

Can you please document that this call does not verify as to whether the document exists or not.

question: If documents cannot be registered multiple times, and doc ids are deterministic content ids, how are you planning on handling multiple nodes registering the same document independently (before each node is aware that the other node has such a document)?

Reply:
> > You can't / should not register the document after it has already been registered
> 
> I'm not talking about multiple registrations here, I'm asking about document creation. I.e. create doc, wait, update doc, wait, etc, etc, register doc.

The Object (in this case document) must register the relation the moment we want to start access control. With Synchronous Register that moment is on creation (when Aync register is implemented the document will have a middle stage where it is in transit to register but currently not registered). In your example if I understand correctly: `create doc, wait, update doc, wait, etc, etc, register doc` the update doc command can be done by anyone even if they don't have access to the document and the register doc happened after create.


> > Are you asking if a public document (unregistered document) can be registered later?
> 
> All documents are unregistered at somepoint with the current flow - as the document must exist before it is registered, and if it is not registered and does exist, it must be unregistered no?

When the request of CreateDocument happens on a permissioned collection, and an identity exists, the document MUST be registered. If it is not registered(i.e. unregistered) and does exist then it is public.

Note: Not all documents are registered on create as we might be missing signature or acp module might be turned off


> > ACP has no knowledge of docID valid or not, is why all the steps are done after document is created, if you look at when it is called.
> 
> Can you please document that this call does not verify as to whether the document exists or not.

This is on the acp module, whose sole purpose is to interact with acp, I don't know why someone would think it is responsible to test document existence and additional validation, all of that happens on the functions that call this call. But since you raised it I will document it.

> question: If documents cannot be registered multiple times, and doc ids are deterministic content ids, how are you planning on handling multiple nodes registering the same document independently (before each node is aware that the other node has such a document)?

It is not the node that is aware of document registration, but that node's acp module, for now every node has it's own acp module, In future the acp module can be shared (as that is how sourcehub is going to use all defradb nodes). The shared modules will sort those cases, also note P2P stuff is out-of-scope and haven't been implemented or tested, this case though definitely will be handled there. 

- [ ] Document that the registering acp method doesn't verify if document exists or not

Reply:
>  With Synchronous Register that moment is on creation

That is only because that is what we currently want to do (outside of this interface) though?  There is nothing within the interface that demands this?

> When the request of CreateDocument happens on a permissioned collection, and an identity exists, the document MUST be registered. If it is not registered(i.e. unregistered) and does exist then it is public.

How do you plan on handling async doc registration if everything is public by default?

Reply:
> > With Synchronous Register that moment is on creation
> 
> That is only because that is what we currently want to do (outside of this interface) though? There is nothing within the interface that demands this?

Correct, Async Registering is on the Roadmap. ACPModule Interface will be used to achieve this, but yes it's not `within` the interface that demands this.


> > When the request of CreateDocument happens on a permissioned collection, and an identity exists, the document MUST be registered. If it is not registered(i.e. unregistered) and does exist then it is public.
> 
> How do you plan on handling async doc registration if everything is public by default?

I am not sure what you mean by `everything is public by default`.

Documents are only public if either one of the following is true:
1) Doc Creation happens, and ACP module is unavailable.
2) Doc Creation happens, and Schema Is Not Permissioned (i.e. no policy).
3) Doc Creation happens, from a request that did not specify any Identity

In all the cases above that is the desired behavior.

> How do you plan on handling async doc registration

This is out of scope for this PR but I can explain a rough implementation.

Async Doc registration would happen only for documents that are created with all 3 conditions met:
1) Have ACP Module
2) Have Signature
3) Have Permissioned Schema

In other words, same as right now. But the only difference is how we handle requests to document when it is `in-transit` to be registered with acp and if it wasn't yet registered.

Then here is what will solve that scenario:
- We will store the `intransit` document which will only be accessible by it's owner (it cant be shared until acp module registration suceeds)
- We have an async events like action that over some period can keep attempting to register that document.
- Once the document is registered then will act as a normal registered doc.
- If at any point acp is turned off while registration is in transit, we can decide how we want to handle that case.

Reply:
> Correct, Async Registering is on the Roadmap. ACPModule Interface will be used to achieve this, but yes it's not within the interface that demands this.

Then why does the interface say that it is a requirement?

> The rest :)

Thanks for the explanation :) I thought you meant that 'in-transit' docs would essentially be public until registration completed.  I am still a bit concerned that they may be 'lost' when in-transit (or if registration fails), but that is an easier thing for me to kick on down the road :)

Reply:
> > Correct, Async Registering is on the Roadmap. ACPModule Interface will be used to achieve this, but yes it's not within the interface that demands this.
> 
> Then why does the interface say that it is a requirement?

Why does it say document should be registered on creation? Because that is a requirement, even when the async register is implemented the event queue would be populated upon failure from a async request that was made at creation of the doc.

However I do see in the future if we want to allow a user to manually specify a public document to be registered later.

I can rename it, as I saw another suggestion for this too.

- [ ] Rename the func that registers document to not include `On Creation`